mempool/v1/reactor.go

```diff
@@ -0,0 +1,394 @@
+// txs to the peers you received it from.
+type Reactor struct {
+	service.BaseService
+
+	config  *cfg.MempoolConfig
+	mempool *TxMempool
+	ids     *mempool.MempoolIDs
+
+	// XXX: Currently, this is the only way to get information about a peer. Ideally,
+	// we rely on message-oriented communication to get necessary peer data.
+	// ref: https://github.com/tendermint/tendermint/issues/5670
+	peerMgr PeerManager
+
+	mempoolCh   *p2p.Channel
+	peerUpdates *p2p.PeerUpdates
+	closeCh     chan struct{}
+
+	// peerWG is used to coordinate graceful termination of all peer broadcasting
+	// goroutines.
+	peerWG sync.WaitGroup
+
+	mtx          tmsync.Mutex
+	peerRoutines map[p2p.NodeID]*tmsync.Closer
+}
+
+// NewReactor returns a reference to a new reactor.
+func NewReactor(
+	logger log.Logger,
+	config *cfg.MempoolConfig,
+	peerMgr PeerManager,
+	txmp *TxMempool,
+	mempoolCh *p2p.Channel,
+	peerUpdates *p2p.PeerUpdates,
+) *Reactor {
+
+	r := &Reactor{
+		config:       config,
+		peerMgr:      peerMgr,
+		mempool:      txmp,
+		ids:          mempool.NewMempoolIDs(),
+		mempoolCh:    mempoolCh,
+		peerUpdates:  peerUpdates,
+		closeCh:      make(chan struct{}),
+		peerRoutines: make(map[p2p.NodeID]*tmsync.Closer),
+	}
+
+	r.BaseService = *service.NewBaseService(logger, "Mempool", r)
+	return r
+}
+
+// GetChannelShims returns a map of ChannelDescriptorShim objects, where each
+// object wraps a reference to a legacy p2p ChannelDescriptor and the corresponding
+// p2p proto.Message the new p2p Channel is responsible for handling.
+//
+//
+// TODO: Remove once p2p refactor is complete.
+// ref: https://github.com/tendermint/tendermint/issues/5670
+func GetChannelShims(config *cfg.MempoolConfig) map[p2p.ChannelID]*p2p.ChannelDescriptorShim {
+	largestTx := make([]byte, config.MaxTxBytes)
+	batchMsg := protomem.Message{
+		Sum: &protomem.Message_Txs{
+			Txs: &protomem.Txs{Txs: [][]byte{largestTx}},
+		},
+	}
+
+	return map[p2p.ChannelID]*p2p.ChannelDescriptorShim{
+		mempool.MempoolChannel: {
+			MsgType: new(protomem.Message),
+			Descriptor: &p2p.ChannelDescriptor{
+				ID:                  byte(mempool.MempoolChannel),
+				Priority:            5,
+				RecvMessageCapacity: batchMsg.Size(),
+
+				MaxSendBytes: 5000,
+			},
+		},
+	}
+}
+
+// OnStart starts separate go routines for each p2p Channel and listens for
+// envelopes on each. In addition, it also listens for peer updates and handles
+// messages on that p2p channel accordingly. The caller must be sure to execute
+// OnStop to ensure the outbound p2p Channels are closed.
+func (r *Reactor) OnStart() error {
+	if !r.config.Broadcast {
+		r.Logger.Info("tx broadcasting is disabled")
+	}
+
+	go r.processMempoolCh()
+	go r.processPeerUpdates()
+
+	return nil
+}
+
+// OnStop stops the reactor by signaling to all spawned goroutines to exit and
+// blocking until they all exit.
+func (r *Reactor) OnStop() {
+	r.mtx.Lock()
+	for _, c := range r.peerRoutines {
+		c.Close()
+	}
+	r.mtx.Unlock()
```

Comment:
we should do this in a different function with defers, in case any of the close methods panic?

Reply:
You mean if `c.Close()` panics? I mean we could, but we know it won't.